### ISSUE TITLE
Attempts to write file in read-only library during upgrade migration

### ISSUE THREAD

**Issue by kg333** [jellyfin/jellyfin#13048]
### This issue respects the following points:

- [x] This is a **bug**, not a question or a configuration issue; Please visit our forum or chat rooms first to troubleshoot with volunteers, before creating a report. The links can be found [here](https://jellyfin.org/contact/).
- [x] This issue is **not** already reported on [GitHub](https://github.com/jellyfin/jellyfin/issues?q=is%3Aopen+is%3Aissue) _(I've searched it)_.
- [x] I'm using an up to date version of Jellyfin Server stable, unstable or master; We generally do not support previous older versions. If possible, please update to the latest version before opening an issue.
- [x] I agree to follow Jellyfin's [Code of Conduct](https://jellyfin.org/docs/general/community-standards.html#code-of-conduct).
- [x] This report addresses only a single issue; If you encounter multiple issues, kindly create separate reports for each one.

### Description of the bug

After installing the latest 10.10.2 package of Jellyfin on Ubuntu 22.04.5 LTS (10.10.2+ubu2204), Jellyfin fails to start with a bug attempting to access an m3u playlist in a read-only library.  Based on the migration description, it appears to be trying to delete it during migration `RemoveDuplicatePlaylistChildren`  - however, this is a read-only library and Jellyfin shouldn't be attempting to modify any files.  Jellyfin is able to read the file via `sudo -u jellyfin cat <path to m3u>`, so it does have correct read-only access.

### Reproduction steps

1. Set up Jellyfin 10.9.11.0 instance with read-only music library containing m3u playlists. (ex. /home/matthew/Music)
2. Install the updated Jellyfin package on Ubuntu 22.04.5 for version 10.10.2.
3. Attempt to start Jellyfin and observe failure to start when it attempts to access m3u files.

### What is the current _bug_ behavior?

Jellyfin fails to complete migration 'RemoveDuplicatePlaylistChildren" while attempting to delete an m3u it doesn't have write access to, and fails to start.

### What is the expected _correct_ behavior?

Jellyfin should not attempt cleanup or deletion of files in read-only libraries and should start.

### Jellyfin Server version

10.10.0+

### Specify commit id

Not known

### Specify unstable release number

Not known

### Specify version number

10.10.2+ubu2204

### Specify the build version

Unable to access dashboard, failure to start

### Environment

```markdown
- OS: Ubuntu 22.04.5 LTS Jammy
- Linux Kernel: 5.15.0-125-generic #135-Ubuntu SMP Fri Sep 27 13:53:58 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
- Virtualization: Proxmox VM
- Clients: N/A
- Browser: N/A
- FFmpeg Version: N/A
- Playback Method: N/A
- Hardware Acceleration: None
- GPU Model: None
- Plugins: None
- Reverse Proxy: Apache
- Base URL: none
- Networking: NFS share
- Storage: NFS
```

### Jellyfin logs

```shell
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [WRN] Unable to find linked item at path ..\Hotline Miami\TurfIntro.ogg
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [WRN] Unable to find linked item at path ..\Hotline Miami\TurfMain.ogg
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [WRN] Unable to find linked item at path ..\Iron Maiden\(2000) Brave New World\01 - The Wickerman.mp3
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [WRN] Unable to find linked item at path ..\Sonata Arctica\SONATA_A.MP3
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [WRN] Unable to find linked item at path ..\Edguy\Edguy - 06 - The Asylum.mp3
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [WRN] Unable to find linked item at path ..\Eisbrecher\Eisbrecher - This Is Deutsch.mp3
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [WRN] Unable to find linked item at path ..\Gamma Ray\Gamma_Ray-Land_Of_The_Free_II-Promo-2007-QTXMp3\07-empress-qtxmp3.mp3
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [WRN] Unable to find linked item at path ..\Gamma Ray\Gamma Ray - Fight.mp3
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [WRN] Unable to find linked item at path ..\Kamelot\08 - The Black Halo - 2005\02 - When The Lights Are Down (Feat. Jens Johannson).mp3
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [WRN] Unable to find linked item at path ..\Within Temptation\THE_SILE\ANGELS__.MP3
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [WRN] Unable to find linked item at path ..\Sonata Arctica\PAID_IN_\SONAT001.MP3
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [ERR] Could not apply migration 'RemoveDuplicatePlaylistChildren'
Nov 16 16:00:44 jellyfin jellyfin[2000]: System.UnauthorizedAccessException: Access to the path '/home/matthew/Music/Playlists/Wakeup - Copy.m3u' is denied.
Nov 16 16:00:44 jellyfin jellyfin[2000]:  ---> System.IO.IOException: Permission denied
Nov 16 16:00:44 jellyfin jellyfin[2000]:    --- End of inner exception stack trace ---
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirError)
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String path, OpenFlags flags, Int32 mode, Boolean failForSymlink, Boolean& wasSymlink, Func`4 createOpenException)
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, UnixFileMode openPermissions, Int64& fileLength, UnixFileMode& filePermissions, Boolean failForSymlink, Boolean& wasSymlink, Func`4 createOpenException)
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at System.IO.File.OpenHandle(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at System.IO.File.WriteToFile(String path, FileMode mode, String contents, Encoding encoding)
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at Emby.Server.Implementations.Playlists.PlaylistManager.SavePlaylistFile(Playlist item)
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at Jellyfin.Server.Migrations.Routines.RemoveDuplicatePlaylistChildren.Perform()
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at Jellyfin.Server.Migrations.MigrationRunner.PerformMigrations(IMigrationRoutine[] migrations, MigrationOptions migrationOptions, Action`1 saveConfiguration, ILogger logger)
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [FTL] Error while starting server
Nov 16 16:00:44 jellyfin jellyfin[2000]: System.UnauthorizedAccessException: Access to the path '/home/matthew/Music/Playlists/Wakeup - Copy.m3u' is denied.
Nov 16 16:00:44 jellyfin jellyfin[2000]:  ---> System.IO.IOException: Permission denied
Nov 16 16:00:44 jellyfin jellyfin[2000]:    --- End of inner exception stack trace ---
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirError)
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String path, OpenFlags flags, Int32 mode, Boolean failForSymlink, Boolean& wasSymlink, Func`4 createOpenException)
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, UnixFileMode openPermissions, Int64& fileLength, UnixFileMode& filePermissions, Boolean failForSymlink, Boolean& wasSymlink, Func`4 createOpenException)
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at System.IO.File.OpenHandle(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at System.IO.File.WriteToFile(String path, FileMode mode, String contents, Encoding encoding)
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at Emby.Server.Implementations.Playlists.PlaylistManager.SavePlaylistFile(Playlist item)
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at Jellyfin.Server.Migrations.Routines.RemoveDuplicatePlaylistChildren.Perform()
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at Jellyfin.Server.Migrations.MigrationRunner.PerformMigrations(IMigrationRoutine[] migrations, MigrationOptions migrationOptions, Action`1 saveConfiguration, ILogger logger)
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at Jellyfin.Server.Migrations.MigrationRunner.Run(CoreAppHost host, ILoggerFactory loggerFactory)
Nov 16 16:00:44 jellyfin jellyfin[2000]:    at Jellyfin.Server.Program.StartServer(IServerApplicationPaths appPaths, StartupOptions options, IConfiguration startupConfig)
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [INF] Running query planner optimizations in the database... This might take a while
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [INF] Disposing CoreAppHost
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [INF] Disposing MusicBrainzArtistProvider
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [INF] Disposing MusicBrainzAlbumProvider
Nov 16 16:00:44 jellyfin jellyfin[2000]: [16:00:44] [INF] Disposing PluginManager
Nov 16 16:00:44 jellyfin systemd[1]: jellyfin.service: Deactivated successfully.
Nov 16 16:00:44 jellyfin systemd[1]: jellyfin.service: Consumed 6.338s CPU time.
```

### FFmpeg logs

```shell
N/A, not transcoding related
```

### Client / Browser logs

N/A, failure to start

### Additional information

Permissions on file in question:
```
-rw-r--r-- 1 matthew matthew    831 Nov 21  2012 'Wakeup - Copy.m3u'
```

I've rolled back to 10.9.11+ubu2204 for now and would be interested in workarounds to disable the `RemoveDuplicatePlaylistChildren` migration.  Can I edit something in `migrations.xml` to force that migration to be skipped?

**Comment by b9AcE** [comment]
I seem to have encountered something similar, but I already restored to my previous state from system-backup before finding this problem report.
Notably however, it kept complaining about some .m3u-file's read-only filesystem even though I in response to earlier complaints had deleted every .m3u (and .m3u8) on the machine, so it couldn't possibly be on a "read-only filesystem" when it didn't exist.

**Comment by solidsnake1298** [comment]
I was able to reproduce.  Both with the docker volume mounted as read only and by setting the individual m3u file's permissions so that the container user had read only permissions.

**Comment by b9AcE** [comment]
Whatever it is trying to migrate seems to not get bypassed, as a workaround, by deleting the playlist-files in question.
I did the same again after already having deleted every *.m3u and *.m3u8 file, but it turned out there was also *.pls and same as before even after deleting the .pls file jellyfin kept complaining that it was read-only while no longer actually existing, so it's presumably something else causing the failure to trigger. Database-related?

**Comment by solidsnake1298** [comment]
I set the permissions to allow writing and the m3u went from this.

```
/media/storage2/testMusic/Smashing Pumpkins/Smashing Pumpkins- Bullet with Butterfly Wings.mp3
/media/storage2/testMusic/Smashing Pumpkins/Smashing Pumpkins - Tonight, Tonight.mp3
```

A simple list with absolute paths to this.

```
#EXTM3U
#EXTALB:The Aeroplane Flies High
#EXTART:The Smashing Pumpkins
#EXTINF:260,Bullet With Butterfly Wings
Smashing Pumpkins- Bullet with Butterfly Wings.mp3
#EXTALB:Tonight, Tonight
#EXTART:The Smashing Pumpkins
#EXTINF:254,Tonight, Tonight
Smashing Pumpkins - Tonight, Tonight.mp3
```

A formatted m3u with relative paths.

**Comment by solidsnake1298** [comment]
> Whatever it is trying to migrate seems to not get bypassed, as a workaround, by deleting the playlist-files in question. I did the same again after already having deleted every *.m3u and *.m3u8 file, but it turned out there was also *.pls and same as before even after deleting the .pls file jellyfin kept complaining that it was read-only while no longer actually existing, so it's presumably something else causing the failure to trigger. Database-related?

I think the better solution than just deleting the playlist files, is to back them up and make them writable.  Assuming you need/want to upgrade to 10.10.2.

**Comment by b9AcE** [comment]
I had my media-directories bind-mounted read-only in my docker-compose file, although I don't remember why and it seems pointless, so I removed that restriction now, as well as removed all playlist-format files I could find just because I don't' use them anyway and then an update went fine.
Don't know which of those were the problem as I did them in parallel, but either or both.

**Comment by farotherside** [comment]
just as another couple of datapoints, I was able to get jellyfin running by adjusting the umask in fstab to _temporarily_ allow the jellyfin user to write to the drive - but even though this updating of m3u files is harmless and well-intentioned, I still think it's a highly problematic stance to take for a project to quietly modify files on a mediaserver without first asking permission from (or even notifying!!) the owner. sketchy behavior, guys.

secondly, the server was not able to come up without further help, because the migration process was referencing missing/inaccessible M3U files _on a drive that I had removed last week_ from the mediaserver. I had removed all mention of that drive from the various libraries, but apparently the "playlists" library/database still had entries for those M3Us.

honestly I don't ever use M3Us and I don't use the 'playlists' library feature, but I have neither come across anything in the settings to configure it nor disable it. at the very least it would seem that whatever database playlists are in is not being updated by nightly library scans.

**Comment by sibbi77** [comment]
As a workaround, you can add

```
<ValueTupleOfGuidString>
  <Item1>96c156a2-7a13-4b3b-a8b8-fb80c94d20c0</Item1>
  <Item2>RemoveDuplicatePlaylistChildren</Item2>
</ValueTupleOfGuidString>
```

to the file `config/migrations.xml`. Jellyfin will then show

`[DBG] [1] Jellyfin.Server.Migrations.MigrationRunner: Skipping migration 'RemoveDuplicatePlaylistChildren' since it is already applied`

in its log.

**Comment by Shadowghost** [comment]
I guess we need to exclude library-based playlists from this check in the migration.

**Comment by Shadowghost** [comment]
If anyone could try with the above linked PR and check if it fixes the issue I'd be grateful.

**Comment by Dnugs** [comment]
Just to say I'm getting the same error with the latest docker image but removing the m3u file in question doesn't seem to resolve the issue.  though getting rid of the read-only option in my .yml file and recreating resolved the issue.

**Comment by harry8326** [comment]
The workaround is working, had the problem like above when I updated my lxc container right now. The workaround fixed it, server is starting now.

**Comment by sgoldthorpe** [comment]
Also had same issue as I mount some of my media shares read-only.  Workaround change to migrations.xml allowed me to bring my server back up.

**Comment by Josef-Friedrich** [comment]
I also had the same issue. Im using the apt version on Ubuntu 24.04. I had problems to find the `migrations.xml` file. For my setup the location is: `/etc/jellyfin/migrations.xml`

**Comment by Shadowghost** [comment]
Fixed by #13059
